Ref #2 add IronMq integration

filosganga opened the pull request:
This PR contains a proposal for IronMQ integration. This implementation only supports "at most one" delivery guarantee. Adding support for "at least one" will be a future work.

filosganga commented:
No idea why it is failing on travis-ci it seems that `MqttSourceSpec` and `MqttSourceTest` fail to bind because the port is in use, but I don't use that port 🤔

patriknw commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,72 @@
+# IronMq Connector
+
+The IronMq connector provides an Akk stream source and sink to connect to the [IronMQ](https://www.iron.io/platform/ironmq/) queue.
```
s/Akk/Akka/

patriknw commented on ironmq/src/main/scala/akka/stream/alpakka/ironmq/IronMqClient.scala:
```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.ironmq
```
We use the convention of `scaladsl` and `javadsl` packages for the Scala and Java APIs

patriknw replied:
Is this user API?

filosganga replied:
It isnot user API It is internal.

patriknw commented on ironmq/src/main/scala/akka/stream/alpakka/ironmq/IronMqClient.scala:
```diff
@@ -0,0 +1,248 @@
+
+  import Codec._
+
+  private implicit val materializer = ActorMaterializer()
```
I think it would be better to pass the `Materializer` as parameter. Where is it used? Should it be param to those methods or to the constructor?

filosganga replied:
It is used for the Akka HTTP. Not sure what is the best approach in this case.  I guess as the actor system is passed in the constructor, the materializer should be passed to the constructor as well.

patriknw commented:
@filosganga ping?

filosganga commented:
@patriknw I am still working on it, I had some problem finding a proper way to pass the materializer to the client. I have found a solution now.

patriknw commented:
thanks for the update

ktoso commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,72 @@
+
+At this time only a at-most-once guarantee consumer is supported (the messages are deleted straight away on consumption). 
+
+The consumer is pull one. It will pull every `n` milliseconds, waiting for `m` milliseconds to consume new messages and
```
The consumer is a pull one?
The consumer is pull based?

filosganga replied:
Pull based yeah

ktoso commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,72 @@
+At this time only a at-most-once guarantee consumer is supported (the messages are deleted straight away on consumption). 
+
+The consumer is pull one. It will pull every `n` milliseconds, waiting for `m` milliseconds to consume new messages and
+will push them to the downstream. These parameters are not configurable at the moment.
```
They definitely should be before we bring it into alpakka though

filosganga replied:
Makes sense

ktoso commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,72 @@
+will push them to the downstream. These parameters are not configurable at the moment.
+
+The consumer could be instantiated using the @scaladoc[IronMqConsumer](akka.stream.alpakka.ironmq.scaladsl.IronMqConsumer). 
+It provides methods to obtain either a `Source[Message, NotUsed]`. You will need a different consumer for each queue, but
```
either a ... or what?

filosganga replied:
Uhm good point, communication is not my best skill

ktoso commented on ironmq/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,13 @@
+    project-id = ${?IRONMQ_PROJECT_ID}
+
+    token = "<fill with ironMq token>"
+    token = ${?IRONMQ_TOKEN}
```
not sure reference should take ENV vars, that's up to the users to configure

filosganga replied:
Agree it smells a bit

ktoso commented on ironmq/src/main/scala/akka/stream/alpakka/ironmq/Codec.scala:
```diff
@@ -0,0 +1,67 @@
+package akka.stream.alpakka.ironmq
+
+import io.circe.{ Decoder, Encoder, Json }
+import io.circe.syntax._
```
Hm... I guess it's ok, we won't be able to stick to a single JSON lib I guess - and circe is nice (though we usually used spray)

ktoso commented on ironmq/src/main/scala/akka/stream/alpakka/ironmq/IronMqClient.scala:
```diff
@@ -0,0 +1,244 @@
+
+  private val pipeline: Flow[HttpRequest, HttpResponse, _] = Flow[HttpRequest]
+    .map(_.withHeaders(Authorization(GenericHttpCredentials("OAuth", settings.token))))
+    .map(_ -> NotUsed)
```
map to not used, what?

filosganga replied:
The syntax is misleading is a map ```x => (x, NotUsed)```

ktoso replied:
lol, right - reading java PRs and then this tricked me :-)

filosganga replied:
I can change it to be less misleading if you think it is necessary.

ktoso replied:
Nah, it's fine - I was just in "java parsing mode" back there

filosganga commented:
I have added the at-least-once semantic and fixed all the comments. I just would like to add the committing feature to the producer to create a close loop and then it should go to go I guess.

filosganga commented:
Need to add some more test coverage for the committable aware producers

filosganga commented:
@patriknw @ktoso It ready to be reviewed again.

filosganga commented:
@patriknw @ktoso ping

patriknw commented on ironmq/src/main/resources/reference.conf:
```diff
@@ -0,0 +1,33 @@
+    reservation-timeout = 30 seconds
+
+  }
+}
```
a newline character is actually important in those conf files, because otherwise it will be wrong when using tools like sbt-assembly that merges those files

filosganga replied:
Fixed

patriknw commented on ironmq/src/test/scala/akka/stream/alpakka/ironmq/AkkaFixture.scala:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  override protected def afterEach(): Unit = {
+    actorSystem.terminate()
```
await also, so it's completely shutdown before next test

filosganga replied:
Fixed

patriknw commented on ironmq/src/main/scala/akka/stream/alpakka/ironmq/CommittableMessage.scala:
```diff
@@ -0,0 +1,22 @@
+ * Commit an offset that is included in a [[CommittableMessage]].
+ */
+trait Committable {
+  def commit(): Future[Done]
```
do we need something else for javadsl?

filosganga replied:
I am expecting people will use the `scala.compat.java8.FutureConverters` to convert it to a `CompletionStage[Done]`.

```
import akka.Done
import java.util.concurrent._
import scala.compat.java8.FutureConverters._

val committable: Committable = new Committable {
  def commit(): Future[Done] = Future(Done)
}

val done: CompletionStage[Done] = committable.commit().asJava
```

I can add two methods `commitJava` and `commitScala` but I really don't like it. In that case, I will prefer having a different `Committable` for Java.

ktoso replied:
We can not force Java people to do extra work, our APIs need to be just as good for both languages.

I suggest doing `ironmq/src/main/scala/akka/stream/alpakka/ironmq/javadsl/CommittableMessage.scala`
and `ironmq/src/main/scala/akka/stream/alpakka/ironmq/scaladsl/CommittableMessage.scala`

as other components in Alpakka and Akka Streams itself does.

You can get away without implementing the stage twice - simply in each javadsl/scaladsl you'd call the same `Graph[...]` (the actual impl).

filosganga replied:
@ktoso It makes sense will do that

filosganga replied:
Done

filosganga commented:
@patriknw @ktoso Ready to review again

ktoso commented on ironmq/src/test/scala/akka/stream/alpakka/ironmq/AkkaFixture.scala:
```diff
@@ -18,7 +27,12 @@ trait AkkaFixture extends ConfigFixture with BeforeAndAfterEach { _: Suite =>
 }
+
+object AkkaFixture {
+  val DefaultActorSystemTerminateTimeout: Duration = 10.seconds
```
OK, though would be good to at some point make sure all Alpakka tests use the same style here... not neccessary in this PR though, is good for now

ktoso commented on ironmq/src/main/resources/reference.conf:
```diff
@@ -30,4 +30,4 @@ akka.stream.alpakka.ironmq {
   }
-}
\ No newline at end of file
+}
```
thanks, that's very important in conf files :)

filosganga commented:
@ktoso @patriknw I have defined a different `Committable` and `CommittableMessage` for Scala and Java DSL. Can you please review it again?

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+IronMq can be used either in cloud or on-premise. Either way you need a authentication token and a project ID. These can
+be set in the Typesafe config:
+
+: @@snip (../../../../ironmq/src/main/resources/reference.conf)
```
Remove the colon before `@@snip ...`, as otherwise the previous paragraph is being treated as a snippet header text.

2m commented on build.sbt:
```diff
@@ -48,6 +48,13 @@ lazy val hbase = project
+    name := "akka-stream-alpakka-ironmq",
+    Dependencies.IronMq
+  )
+
```
Also add `ironmq` to the root `alpakka` project at the top of this file. Otherwise no ironmq connector tests are ran by Travis.

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+The consumer is poll based one. It will poll every `n` milliseconds, waiting for `m` milliseconds to consume new messages and
+will push them to the downstream. All These parameters are configurable by the Typesafe config.
+
+It support both at-most-once and at-least-once semantic. In the first case the messages are deleted straight away after 
```
... It support*s* ... semantic*s* ...

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+been fetched. In the latter case the messages piggy back a Committable object that should be used to commit the message.
+Committing the message will cause the message to be deleted from the queue.
+
+The consumer could be instantiated using the @scaladoc[IronMqConsumer](akka.stream.alpakka.ironmq.scaladsl.IronMqConsumer). 
```
Missing $ at the end of the classpath as this is an object.

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+for at-most-one semantic, the latter for at-least-once semantic. 
+
+### Producer
+The producer is very trivial at this time, it does not provide any batching mechanism., but send messages to IronMq as 
```
Remove full stop before comma. Also ... but send*s* messages ...

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+The producer is very trivial at this time, it does not provide any batching mechanism., but send messages to IronMq as 
+soon as they arrive to the stage.
+
+The producer could be instantiated using the @scaladoc[IronMqProducer](akka.stream.alpakka.ironmq.scaladsl.IronMqProducer).
```
Also add $ at the end of the class path.

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+The producer could be instantiated using the @scaladoc[IronMqProducer](akka.stream.alpakka.ironmq.scaladsl.IronMqProducer).
+It provides methods to obtain either a `Flow[PushMessage, Messages.Id, NotUsed]` or a `Sink[PushMessage, NotUsed]`. 
+
+The @scaladoc[PushMessage](akka.stream.alpakka.ironmq.PushMessage) allow to specify the delay per individual message. The
```
... allows*s* ...

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+It provides methods to obtain either a `Flow[PushMessage, Messages.Id, NotUsed]` or a `Sink[PushMessage, NotUsed]`. 
+
+The @scaladoc[PushMessage](akka.stream.alpakka.ironmq.PushMessage) allow to specify the delay per individual message. The
+message expiration is set a queue level. The @scaladoc[IronMqClient](akka.stream.alpakka.ironmq.IronMqClient) does not 
```
What do you mean ... is set a queue level? Does the message expiration set a queue level? Or is message expiration set per queue?

filosganga replied:
The message expiration is set per queue and cannot be changed for a singular message.

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+message expiration is set a queue level. The @scaladoc[IronMqClient](akka.stream.alpakka.ironmq.IronMqClient) does not 
+still have all the features to create and update these settings on a queue.
+
+If you are using the producer `Flow` The returned @scaladoc[Messages.Ids](akka.stream.alpakka.ironmq.Messages.Id) will 
```
Class path should be `akka.stream.alpakka.ironmq.Message$$Id`. Change to: `... returned @scaladoc[Message.Id](akka.stream.alpakka.ironmq.Message$$Id)s will ...`

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+If you are using the producer `Flow` The returned @scaladoc[Messages.Ids](akka.stream.alpakka.ironmq.Messages.Id) will 
+contain the ID of the pushed message, that can be used to manipulate the message.
+
+For each `PushMessage` from the upstream you will have exact one `Message.Id` in downstream in the same order. Regardless
```
... exact*ly*

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+For each `PushMessage` from the upstream you will have exact one `Message.Id` in downstream in the same order. Regardless
+if the producer will implement a batch mechanism in the future.
+
+The producer also provide a Committable aware Flow/Sink as `Flow[(PushMessage, ToCommit), (Message.Id, CommitResult), CommitMat]`. 
```
... also provide*s* a ...

2m commented on docs/src/main/paradox/ironmq.md:
```diff
@@ -0,0 +1,76 @@
+if the producer will implement a batch mechanism in the future.
+
+The producer also provide a Committable aware Flow/Sink as `Flow[(PushMessage, ToCommit), (Message.Id, CommitResult), CommitMat]`. 
+It can be used to consume a Flow from an IronMq consumer or any other source that provide a commit mechanism.
```
... that provide*s* a commit ...

2m commented on ironmq/src/main/scala/akka/stream/alpakka/ironmq/javadsl/CommittableMessage.scala:
```diff
@@ -0,0 +1,23 @@
+ */
+trait CommittableMessage extends Committable {
+  def message: Message
+}
```
It would be great to have a simple test written in Java that would validate if the Java Dsl is alright.

2m commented on .travis.yml:
```diff
@@ -38,6 +38,8 @@ deploy:
 env:
   global:
+    - IRONMQ_ENDPOINT=http://localhost:8080
+    - IRONMQ_TOKEN=adminToken
```
Is this really needed? If these are default values, they can be removed.

filosganga commented:
Thanks @2m I did not get a third person right 🥇

2m commented:
No worries. I also always miss articles in my written English. :)

Now I see that scalafmt formatter kicked in when you added ironmq submodule to the aggregate. Travis is complaining now, as it found unformatted code. Run `test:compile` and commit and push reformatted code.

filosganga commented:
Yep I have saw that. I am also trying to write a java test with some difficulties.

filosganga commented:
I have added a very trivial Java test. It causes me to realize how much bad was using the Java DSL. So I have refactored it a bit to look nicer in Java.

2m commented:
Cool. Validation failure was: https://github.com/akka/alpakka/issues/267

Thanks for pushing this one through. :)
